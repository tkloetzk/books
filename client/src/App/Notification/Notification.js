import React from 'react';
import Button from '@material-ui/core/Button';
import CheckCircleIcon from '@material-ui/icons/CheckCircle';
import ErrorIcon from '@material-ui/icons/Error';
import InfoIcon from '@material-ui/icons/Info';
import CloseIcon from '@material-ui/icons/Close';
import green from '@material-ui/core/colors/green';
import amber from '@material-ui/core/colors/amber';
import IconButton from '@material-ui/core/IconButton';
import Snackbar from '@material-ui/core/Snackbar';
import SnackbarContent from '@material-ui/core/SnackbarContent';
import WarningIcon from '@material-ui/icons/Warning';
import { withStyles } from '@material-ui/core/styles';

const variantIcon = {
  success: CheckCircleIcon,
  warning: WarningIcon,
  error: ErrorIcon,
  info: InfoIcon,
};
const styles = () => ({
  success: {
    backgroundColor: green[600],
  },
  errored: {
    backgroundColor: 'red',
  },
  info: {
    backgroundColor: 'yellow',
  },
  warning: {
    backgroundColor: amber[700],
  },
  icon: {
    fontSize: 20,
    marginRight: 5,
  },
  message: {
    display: 'flex',
    alignItems: 'center',
  },
});

const Notification = ({
  classes,
  open,
  handleClose,
  autoHideDuration,
  message,
  type,
}) => {
  if (!message || !type) return null;

  const Icon = variantIcon[type];

  return (
    <Snackbar
      anchorOrigin={{
        vertical: 'bottom',
        horizontal: 'left',
      }}
      open={open}
      autoHideDuration={autoHideDuration}
      onClose={handleClose}
    >
      <SnackbarContent
        className={classes[type]}
        message={
          <span className={classes.message}>
            <Icon className={classes.icon} />
            {message}
          </span>
        }
        action={[
          <IconButton
            key="close"
            aria-label="Close"
            color="inherit"
            className={classes.close}
            onClick={handleClose}
          >
            <CloseIcon className={classes.icon} />
          </IconButton>,
        ]}
      />
    </Snackbar>
  );
};

export default withStyles(styles)(Notification);
