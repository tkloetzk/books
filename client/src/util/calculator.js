import get from 'lodash/get';
import forEach from 'lodash/forEach'
import clone from 'lodash/clone'

let meanGoodreadsVotes
let meanAmazonVotes
let totalMean

export default function sortBooklist(originalBooklist) {
  const booklist = clone(originalBooklist);
  calculateMeans(booklist);
  calculateAdjustedRating(booklist);
  booklist.sort((a, b) =>
    a.adjustedRating < b.adjustedRating
      ? 1
      : b.adjustedRating < a.adjustedRating
      ? -1
      : 0
  );
  return booklist;
}

function calculateMeans(booklist) {
  let goodreadsVotes = 0,
    amazonVotes = 0,
    total = 0;

  forEach(booklist, book => {
    goodreadsVotes += get(book, 'goodreadsAverageRating')
    amazonVotes += get(book, 'amazonAverageRating');
    total +=
      get(book, 'goodreadsAverageRating') +
      get(book, 'amazonAverageRating');
  })

  meanGoodreadsVotes = goodreadsVotes / booklist.length;
  meanAmazonVotes = amazonVotes / booklist.length;
  totalMean = total / 2 / booklist.length;

  return booklist;
}

function calculateAdjustedRating(booklist) {
  for (var i = 0; i < booklist.length; i++) {
    var book = booklist[i];
    var adjustedRating =
      (getAdjustedRating(
        get(book, 'goodreadsRatingsCount'),
        get(book, 'goodreadsAverageRating'),
        meanGoodreadsVotes,
        trimmean(getGoodreadsRatingsCountList(booklist))
      ) +
        getAdjustedRating(
          get(book, 'amazonRatingsCount'),
          get(book, 'amazonAverageRating'),
          meanAmazonVotes,
          trimmean(getAmazonRatingsCountList(booklist))
        )) /
      2;
    book.adjustedRating = adjustedRating;
  }
}

function getAdjustedRating(ratingsCount, averageRating, meanVote, minVotes) {
  //double rating = averageRating * 2;
  return (
    (ratingsCount / (ratingsCount + minVotes)) * averageRating +
    (minVotes / (ratingsCount + minVotes)) * meanVote
  );
}

function getGoodreadsRatingsCountList(booklist) {
  let votes = [];
  for (const book of booklist) {
    votes.push(book.goodreadsRatingsCount);
  }

  return votes.sort(function(a, b) {
    return a - b;
  });
}

function getAmazonRatingsCountList(booklist) {
  let votes = [];
  for (const book of booklist) {
    votes.push(book.amazonRatingsCount);
  }

  return votes.sort(function(a, b) {
    return a - b;
  });
}

function trimmean(votes) {
  votes.sort(function(a, b) {
    return a - b;
  });

  const removeAmount = (votes.length * 0.25) / 2;
  const roundDownNearedMultipleTwo = Math.floor(
    removeAmount >= 0
      ? (removeAmount / 2) * 2
      : ((removeAmount - 2 + 1) / 2) * 2
  );
  return getMean(
    votes.slice(
      roundDownNearedMultipleTwo,
      votes.length - roundDownNearedMultipleTwo
    )
  );
}

function getMean(array) {
  let sum = 0;
  for (let i = 0; i < array.length; i++) {
    sum += array[i];
  }
  return sum / array.length;
}
