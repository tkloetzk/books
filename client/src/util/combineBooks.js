import forEach from 'lodash/forEach';
import mergeByKey from 'array-merge-by-key';
import isEmpty from 'lodash/isEmpty';

export function combineBooks(
  amazonBooks,
  googleBooks,
  goodreadsBooks,
  bookshelf
) {
  const combinedBooks = mergeByKey(
    'isbn',
    amazonBooks,
    googleBooks,
    goodreadsBooks
  );
  let duplicates = [];
  let duplicatedISBNs = [];
  forEach(combinedBooks, duplicatedBook => {
    return forEach(bookshelf, existingBook => {
      if (duplicatedBook.isbn === existingBook.isbn) {
        duplicatedBook.differences = this.compareDifferences(
          existingBook,
          duplicatedBook,
          []
        );
        // TODO: If duplicate but no differences exist, don't add but show notification
        if (duplicatedBook.differences.length) {
          duplicatedBook._id = existingBook._id;
          duplicates.push(duplicatedBook);
        } else {
          duplicatedISBNs.push({ isbn: duplicatedBook.isbn });
        }
      }
    });
  });
  return { combinedBooks, duplicates, duplicatedISBNs };
}
export function compareDifferences(oldBook, newBook, difference) {
  Object.keys(oldBook).forEach(key => {
    if (typeof oldBook[key] !== 'object') {
      if (
        oldBook[key] !== newBook[key] &&
        key !== '__v' &&
        key !== '_id' &&
        key !== 'adjustedRating' &&
        newBook[key] !== ''
      )
        difference.push({
          key,
          currentValue: oldBook[key],
          newValue: newBook[key],
        });
    } else {
      compareDifferences(oldBook[key], newBook[key], difference);
    }
  }, difference);

  return difference;
}

export default { combineBooks, compareDifferences };
