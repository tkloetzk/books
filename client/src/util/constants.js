export const LOADING_STATUSES = {
  initial: 'initial',
  loading: 'loading',
  success: 'success',
  errored: 'errored',
};
