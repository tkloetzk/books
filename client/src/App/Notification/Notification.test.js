import React from 'react';
import Notification from './Notification';
import { mount } from 'enzyme';
import { LOADING_STATUSES } from '../../util/constants';

describe('Notification', () => {
  let props;
  let wrapper;

  beforeEach(() => {
    props = {
      classes: {},
      open: false,
      handleClose: jest.fn(),
      autoHideDuration: 4000,
      message: '',
      type: LOADING_STATUSES.initial,
    };
    wrapper = mount(<Notification {...props} />);
  });
  afterEach(() => {
    jest.clearAllMocks();
  });
  it('renders null with no props', () => {
    const noPropsWrapper = mount(<Notification />);
    expect(noPropsWrapper).toMatchSnapshot();
  });
  it('render null with no message', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('render null with no type', () => {
    props = Object.assign({ message: 'Message', type: null });
    wrapper = mount(<Notification {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('render with correct success props', () => {
    props = Object.assign({
      message: 'Success',
      open: true,
      type: LOADING_STATUSES.success,
    });
    wrapper = mount(<Notification {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('render with correct success props', () => {
    props = Object.assign({
      message: 'Errored',
      open: true,
      type: LOADING_STATUSES.errored,
    });
    wrapper = mount(<Notification {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
