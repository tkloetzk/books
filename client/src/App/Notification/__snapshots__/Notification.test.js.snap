// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification render null with no message 1`] = `
<Notification
  autoHideDuration={4000}
  classes={
    Object {
      "errored": "Notification-errored-3",
      "icon": "Notification-icon-7",
      "info": "Notification-info-5",
      "initial": "Notification-initial-1",
      "loading": "Notification-loading-4",
      "message": "Notification-message-8",
      "success": "Notification-success-2",
      "undefined": "Notification-undefined-6",
    }
  }
  handleClose={[MockFunction]}
  message=""
  open={false}
  type="initial"
/>
`;

exports[`Notification render null with no type 1`] = `
<Notification
  classes={
    Object {
      "errored": "Notification-errored-3",
      "icon": "Notification-icon-7",
      "info": "Notification-info-5",
      "initial": "Notification-initial-1",
      "loading": "Notification-loading-4",
      "message": "Notification-message-8",
      "success": "Notification-success-2",
      "undefined": "Notification-undefined-6",
    }
  }
  message="Message"
  type={null}
/>
`;

exports[`Notification render with correct errored props 1`] = `"<div class=\\"MuiSnackbar-root-9 MuiSnackbar-anchorOriginBottomLeft-15\\"><div class=\\"MuiTypography-root-46 MuiTypography-body1-55 MuiPaper-root-19 MuiPaper-elevation6-27 MuiSnackbarContent-root-16 Notification-errored-3\\" role=\\"alertdialog\\"><div class=\\"MuiSnackbarContent-message-17\\"><span class=\\"Notification-message-8\\"><svg class=\\"MuiSvgIcon-root-82 Notification-icon-7\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path fill=\\"none\\" d=\\"M0 0h24v24H0z\\"></path><path d=\\"M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z\\"></path></svg>errored</span></div><div class=\\"MuiSnackbarContent-action-18\\"><button class=\\"MuiButtonBase-root-97 MuiIconButton-root-91 MuiIconButton-colorInherit-92\\" tabindex=\\"0\\" type=\\"button\\" aria-label=\\"Close\\"><span class=\\"MuiIconButton-label-96\\"><svg class=\\"MuiSvgIcon-root-82 Notification-icon-7\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z\\"></path><path fill=\\"none\\" d=\\"M0 0h24v24H0z\\"></path></svg></span></button></div></div></div>"`;

exports[`Notification render with correct info props 1`] = `"<div class=\\"MuiSnackbar-root-9 MuiSnackbar-anchorOriginBottomLeft-15\\"><div class=\\"MuiTypography-root-46 MuiTypography-body1-55 MuiPaper-root-19 MuiPaper-elevation6-27 MuiSnackbarContent-root-16 Notification-info-5\\" role=\\"alertdialog\\"><div class=\\"MuiSnackbarContent-message-17\\"><span class=\\"Notification-message-8\\"><svg class=\\"MuiSvgIcon-root-82 Notification-icon-7\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path fill=\\"none\\" d=\\"M0 0h24v24H0z\\"></path><path d=\\"M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z\\"></path></svg>info</span></div><div class=\\"MuiSnackbarContent-action-18\\"><button class=\\"MuiButtonBase-root-97 MuiIconButton-root-91 MuiIconButton-colorInherit-92\\" tabindex=\\"0\\" type=\\"button\\" aria-label=\\"Close\\"><span class=\\"MuiIconButton-label-96\\"><svg class=\\"MuiSvgIcon-root-82 Notification-icon-7\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z\\"></path><path fill=\\"none\\" d=\\"M0 0h24v24H0z\\"></path></svg></span></button></div></div></div>"`;

exports[`Notification render with correct initial props 1`] = `"<div class=\\"MuiSnackbar-root-9 MuiSnackbar-anchorOriginBottomLeft-15\\"><div class=\\"MuiTypography-root-46 MuiTypography-body1-55 MuiPaper-root-19 MuiPaper-elevation6-27 MuiSnackbarContent-root-16 Notification-initial-1\\" role=\\"alertdialog\\"><div class=\\"MuiSnackbarContent-message-17\\"><span class=\\"Notification-message-8\\"><svg class=\\"MuiSvgIcon-root-82 Notification-icon-7\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z\\"></path><path fill=\\"none\\" d=\\"M0 0h24v24H0z\\"></path></svg>initial</span></div><div class=\\"MuiSnackbarContent-action-18\\"><button class=\\"MuiButtonBase-root-97 MuiIconButton-root-91 MuiIconButton-colorInherit-92\\" tabindex=\\"0\\" type=\\"button\\" aria-label=\\"Close\\"><span class=\\"MuiIconButton-label-96\\"><svg class=\\"MuiSvgIcon-root-82 Notification-icon-7\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z\\"></path><path fill=\\"none\\" d=\\"M0 0h24v24H0z\\"></path></svg></span></button></div></div></div>"`;

exports[`Notification render with correct loading props 1`] = `"<div class=\\"MuiSnackbar-root-9 MuiSnackbar-anchorOriginBottomLeft-15\\"><div class=\\"MuiTypography-root-46 MuiTypography-body1-55 MuiPaper-root-19 MuiPaper-elevation6-27 MuiSnackbarContent-root-16 Notification-loading-4\\" role=\\"alertdialog\\"><div class=\\"MuiSnackbarContent-message-17\\"><span class=\\"Notification-message-8\\"><div class=\\"MuiCircularProgress-root-100 MuiCircularProgress-colorPrimary-103 MuiCircularProgress-indeterminate-102 Notification-icon-7\\" style=\\"width:40px;height:40px\\" role=\\"progressbar\\"><svg class=\\"MuiCircularProgress-svg-105\\" viewBox=\\"22 22 44 44\\"><circle class=\\"MuiCircularProgress-circle-106 MuiCircularProgress-circleIndeterminate-108\\" cx=\\"44\\" cy=\\"44\\" r=\\"20.2\\" fill=\\"none\\" stroke-width=\\"3.6\\"></circle></svg></div>loading</span></div><div class=\\"MuiSnackbarContent-action-18\\"><button class=\\"MuiButtonBase-root-97 MuiIconButton-root-91 MuiIconButton-colorInherit-92\\" tabindex=\\"0\\" type=\\"button\\" aria-label=\\"Close\\"><span class=\\"MuiIconButton-label-96\\"><svg class=\\"MuiSvgIcon-root-82 Notification-icon-7\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z\\"></path><path fill=\\"none\\" d=\\"M0 0h24v24H0z\\"></path></svg></span></button></div></div></div>"`;

exports[`Notification render with correct success props 1`] = `"<div class=\\"MuiSnackbar-root-9 MuiSnackbar-anchorOriginBottomLeft-15\\"><div class=\\"MuiTypography-root-46 MuiTypography-body1-55 MuiPaper-root-19 MuiPaper-elevation6-27 MuiSnackbarContent-root-16 Notification-success-2\\" role=\\"alertdialog\\"><div class=\\"MuiSnackbarContent-message-17\\"><span class=\\"Notification-message-8\\"><svg class=\\"MuiSvgIcon-root-82 Notification-icon-7\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path fill=\\"none\\" d=\\"M0 0h24v24H0z\\"></path><path d=\\"M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z\\"></path></svg>success</span></div><div class=\\"MuiSnackbarContent-action-18\\"><button class=\\"MuiButtonBase-root-97 MuiIconButton-root-91 MuiIconButton-colorInherit-92\\" tabindex=\\"0\\" type=\\"button\\" aria-label=\\"Close\\"><span class=\\"MuiIconButton-label-96\\"><svg class=\\"MuiSvgIcon-root-82 Notification-icon-7\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z\\"></path><path fill=\\"none\\" d=\\"M0 0h24v24H0z\\"></path></svg></span></button></div></div></div>"`;

exports[`Notification renders null with no props 1`] = `
<Notification
  classes={
    Object {
      "errored": "Notification-errored-3",
      "icon": "Notification-icon-7",
      "info": "Notification-info-5",
      "initial": "Notification-initial-1",
      "loading": "Notification-loading-4",
      "message": "Notification-message-8",
      "success": "Notification-success-2",
      "undefined": "Notification-undefined-6",
    }
  }
/>
`;
